File: c_sharp/lesson05/lesson05.cs
/* Lesson 05 of NeHe Productions in C#
   created by Sabine Felsinger*/

using System;
using System.Windows.Forms;
using System.Drawing;
using CsGL.OpenGL;

namespace lesson05
{
	public class OurView : OpenGLControl
	{
		public float rtri;			// rtri is for rotating the pyramid
		public float rquad;			// rquad is for rotating the quad

		public bool finished;

		public OurView() : base()
		{
			this.KeyDown += new KeyEventHandler(OurView_OnKeyDown);
			finished = false;
		}
		
		protected void OurView_OnKeyDown(object Sender, KeyEventArgs kea)
		{
			//if escape was pressed exit the application
			if (kea.KeyCode == Keys.Escape) 
			{
				finished = true;
			}
		}

		public override void glDraw()
		{
			GL.glClear(GL.GL_COLOR_BUFFER_BIT | GL.GL_DEPTH_BUFFER_BIT);	// Clear the Screen and the Depth Buffer
			GL.glMatrixMode(GL.GL_MODELVIEW);		// Modelview Matrix
			GL.glLoadIdentity();					// reset the current modelview matrix
			GL.glTranslatef(-1.5f,0.0f,-6.0f);		// move 1.5 Units left and 6 Units into the screen
			GL.glRotatef(rtri,0.0f,1.0f,0.0f);		// rotate the Pyramid on it's Y-axis
			rtri+=0.2f;								// rotation angle

			GL.glBegin(GL.GL_TRIANGLES);			// start drawing a triangle, always counterclockside (top-left-right)
			GL.glColor3f(1.0f,0.0f,0.0f);			// Red
			GL.glVertex3f(0.0f,1.0f,0.0f);			// Top of Triangle (Front)
			GL.glColor3f(0.0f,1.0f,0.0f);			// green
			GL.glVertex3f(-1.0f,-1.0f,1.0f);		// left of Triangle (front)
			GL.glColor3f(0.0f,0.0f,1.0f);			// blue
			GL.glVertex3f(1.0f,-1.0f,1.0f);			// right of triangle (front)

			GL.glColor3f(1.0f,0.0f,0.0f);			// red
			GL.glVertex3f(0.0f,1.0f,0.0f);			// top of triangle (right)
			GL.glColor3f(0.0f,0.0f,1.0f);			// blue
			GL.glVertex3f(1.0f,-1.0f,1.0f);			// left of triangle (right)
			GL.glColor3f(0.0f,1.0f,0.0f);			// green
			GL.glVertex3f(1.0f,-1.0f,-1.0f);		// right of triangel (right)

			GL.glColor3f(1.0f,0.0f,0.0f);			// red
			GL.glVertex3f(0.0f,1.0f,0.0f);			// top of triangle (back)
			GL.glColor3f(0.0f,1.0f,0.0f);			// green
			GL.glVertex3f(1.0f,-1.0f,-1.0f);		// left of triangle (back)
			GL.glColor3f(0.0f,0.0f,1.0f);			// blue
			GL.glVertex3f(-1.0f,-1.0f,-1.0f);		// right of triangle (back)

			GL.glColor3f(1.0f,0.0f,0.0f);			// red
			GL.glVertex3f(0.0f,1.0f,0.0f);			// top of triangle (left)
			GL.glColor3f(0.0f,0.0f,1.0f);			// blue
			GL.glVertex3f(-1.0f,-1.0f,-1.0f);		// left of triangle (left)
			GL.glColor3f(0.0f,1.0f,0.0f);			// green
			GL.glVertex3f(-1.0f,-1.0f,1.0f);		// right of triangle (left)
			GL.glEnd();

			GL.glLoadIdentity();					// reset the current modelview matrix
            GL.glTranslatef(1.5f,0.0f,-7.0f);		// move 1.5 Units right and 7 into the screen
			GL.glRotatef(rquad,1.0f,1.0f,1.0f);		// rotate the quad on the X,Y and Z-axis
			rquad-=0.15f;							// rotation angle

			GL.glBegin(GL.GL_QUADS);				// start drawing a quad
			GL.glColor3f(0.0f,1.0f,0.0f);			// green top
			GL.glVertex3f(1.0f,1.0f,-1.0f);			// top right (top)
            GL.glVertex3f(-1.0f,1.0f,-1.0f);		// top left (top)
			GL.glVertex3f(-1.0f,1.0f,1.0f);			// bottom left (top)
			GL.glVertex3f(1.0f,1.0f,1.0f);			// bottom right (top)

			GL.glColor3f(1.0f,0.5f,0.0f);			// orange
			GL.glVertex3f(1.0f,-1.0f,1.0f);			// top right (bottom)
			GL.glVertex3f(-1.0f,-1.0f,1.0f);		// top left (bottom)
			GL.glVertex3f(-1.0f,-1.0f,-1.0f);		// bottom left (bottom)
			GL.glVertex3f(1.0f,-1.0f,-1.0f);		// bottom right (bottom)

			GL.glColor3f(1.0f,0.0f,0.0f);			// red
			GL.glVertex3f(1.0f,1.0f,1.0f);			// top right (front)
			GL.glVertex3f(-1.0f,1.0f,1.0f);			// top left (front)
			GL.glVertex3f(-1.0f,-1.0f,1.0f);		// bottom left (front)
			GL.glVertex3f(1.0f,-1.0f,1.0f);			// bottom right (front)

			GL.glColor3f(1.0f,1.0f,0.0f);				// yellow
			GL.glVertex3f(-1.0f,1.0f,-1.0f);		// top right (back)
			GL.glVertex3f(1.0f,1.0f,-1.0f);			// top left (back)
			GL.glVertex3f(1.0f,-1.0f,-1.0f);		// bottom left (back)
			GL.glVertex3f(-1.0f,-1.0f,-1.0f);		// bottom right (back)
	
			GL.glColor3f(0.0f,0.0f,1.0f);			// blue
			GL.glVertex3f(-1.0f,1.0f,1.0f);			// top right (left)
			GL.glVertex3f(-1.0f,1.0f,-1.0f);		// top left (left)
			GL.glVertex3f(-1.0f,-1.0f,-1.0f);		// bottom left (left)
			GL.glVertex3f(-1.0f,-1.0f,1.0f);		// bottom right (left)

			GL.glColor3f(1.0f,0.0f,1.0f);			// violett
			GL.glVertex3f(1.0f,1.0f,-1.0f);			// top right (right)
			GL.glVertex3f(1.0f,1.0f,1.0f);			// top left (right)
			GL.glVertex3f(1.0f,-1.0f, 1.0f);		// bottom left (right)
			GL.glVertex3f(1.0f,-1.0f,-1.0f);		// bottom right (right)
			GL.glEnd();

		}

		protected override void InitGLContext() 
		{
			GL.glShadeModel(GL.GL_SMOOTH);								// enable smooth shading
			GL.glClearColor(0.0f, 0.0f, 0.0f, 0.5f);					// black background
			GL.glClearDepth(1.0f);										// depth buffer setup
			GL.glEnable(GL.GL_DEPTH_TEST);								// enables depth testing
			GL.glDepthFunc(GL.GL_LEQUAL);								// type of depth test
			GL.glHint(GL.GL_PERSPECTIVE_CORRECTION_HINT, GL.GL_NICEST);	// nice perspective calculations
			//rtri = 30.0f;			define the rotation angle in the start position of the triangle
			//rquad = 30.0f;		define the rotation angle in the start position of the quad
		}

		protected override void OnSizeChanged(EventArgs e)
		{
			base.OnSizeChanged(e);
			Size s = Size;

			GL.glMatrixMode(GL.GL_PROJECTION);
			GL.glLoadIdentity();
			GL.gluPerspective(45.0f, (double)s.Width /(double) s.Height, 0.1f, 100.0f);	
			GL.glMatrixMode(GL.GL_MODELVIEW);
			GL.glLoadIdentity();
		}
	}
	
	public class MainForm : System.Windows.Forms.Form	
	{
		private lesson05.OurView view;

		public MainForm()
		{
			this.AutoScaleBaseSize = new System.Drawing.Size(5, 13);
			this.ClientSize = new System.Drawing.Size(640, 480);
			this.Name = "MainForm";
			this.Text = "NeHe lesson 05 in C# (by Sabine Felsinger)";
			this.view = new lesson05.OurView();			
			this.view.Parent = this;
			this.view.Dock = DockStyle.Fill; // Will fill whole form
			this.Show();
		}

		static void Main() 
		{
			MainForm form = new MainForm();

			while ((!form.view.finished) && (!form.IsDisposed))		// refreshing the window, so it rotates
			{
				form.view.glDraw();
				form.Refresh();
				Application.DoEvents();
			}

			form.Dispose();
		}
	}
}

File: lwjgl/lesson05/Lesson05/Lesson05.java
/*
 *      This Code Was Created By Jeff Molofee 2000
 *      A HUGE Thanks To Fredric Echols For Cleaning Up
 *      And Optimizing The Base Code, Making It More Flexible!
 *      If You've Found This Code Useful, Please Let Me Know.
 *      Visit My Site At nehe.gamedev.net
 */

import org.lwjgl.opengl.Display;
import org.lwjgl.opengl.DisplayMode;
import org.lwjgl.opengl.GL11;
import org.lwjgl.opengl.glu.GLU;
import org.lwjgl.input.Keyboard;

/**
 * @author Mark Bernard
 * date:    16-Nov-2003
 *
 * Port of NeHe's Lesson 5 to LWJGL
 * Title: 3D Shapes
 * Uses version 0.8alpha of LWJGL http://www.lwjgl.org/
 *
 * Be sure that the LWJGL libraries are in your classpath
 *
 * Ported directly from the C++ version
 *
 * 2004-05-08: Updated to version 0.9alpha of LWJGL.
 *             Changed from all static to all instance objects.
 * 2004-09-22: Updated to version 0.92alpha of LWJGL.
 */
public class Lesson05 {
    private boolean done = false;
    private boolean fullscreen = false;
    private final String windowTitle = "NeHe's OpenGL Lesson 5 for LWJGL (3D Shapes)";
    private boolean f1 = false;

    private float rtri;                 // Angle For The Triangle ( NEW )
    private float rquad;                // Angle For The Quad     ( NEW )
    private DisplayMode displayMode;

    public static void main(String args[]) {
        boolean fullscreen = false;
        if(args.length>0) {
            if(args[0].equalsIgnoreCase("fullscreen")) {
                fullscreen = true;
            }
        }

        Lesson05 l5 = new Lesson05();
        l5.run(fullscreen);
    }
    public void run(boolean fullscreen) {
        this.fullscreen = fullscreen;
        try {
            init();
            while (!done) {
                mainloop();
                render();
                Display.update();
            }
            cleanup();
        }
        catch (Exception e) {
            e.printStackTrace();
            System.exit(0);
        }
    }
    private void mainloop() {
        if(Keyboard.isKeyDown(Keyboard.KEY_ESCAPE)) {       // Exit if Escape is pressed
            done = true;
        }
        if(Display.isCloseRequested()) {                     // Exit if window is closed
            done = true;
        }
        if(Keyboard.isKeyDown(Keyboard.KEY_F1) && !f1) {    // Is F1 Being Pressed?
            f1 = true;                                      // Tell Program F1 Is Being Held
            switchMode();                                   // Toggle Fullscreen / Windowed Mode
        }
        if(!Keyboard.isKeyDown(Keyboard.KEY_F1)) {          // Is F1 Being Pressed?
            f1 = false;
        }
    }

    private void switchMode() {
        fullscreen = !fullscreen;
        try {
            Display.setFullscreen(fullscreen);
        }
        catch(Exception e) {
            e.printStackTrace();
        }
    }

    private boolean render() {
        GL11.glClear(GL11.GL_COLOR_BUFFER_BIT | GL11.GL_DEPTH_BUFFER_BIT);          // Clear The Screen And The Depth Buffer

        GL11.glLoadIdentity();                          // Reset The Current Modelview Matrix

        GL11.glTranslatef(-1.5f,0.0f,-6.0f);                // Move Left 1.5 Units And Into The Screen 6.0
        GL11.glRotatef(rtri,0.0f,1.0f,0.0f);                // Rotate The Triangle On The Y axis ( NEW )
        GL11.glBegin(GL11.GL_TRIANGLES);                    // Drawing Using Triangles
            GL11.glColor3f(1.0f,0.0f,0.0f);             // Red
            GL11.glVertex3f( 0.0f, 1.0f, 0.0f);         // Top Of Triangle (Front)
            GL11.glColor3f(0.0f,1.0f,0.0f);             // Green
            GL11.glVertex3f(-1.0f,-1.0f, 1.0f);         // Left Of Triangle (Front)
            GL11.glColor3f(0.0f,0.0f,1.0f);             // Blue
            GL11.glVertex3f( 1.0f,-1.0f, 1.0f);         // Right Of Triangle (Front)
            GL11.glColor3f(1.0f,0.0f,0.0f);             // Red
            GL11.glVertex3f( 0.0f, 1.0f, 0.0f);         // Top Of Triangle (Right)
            GL11.glColor3f(0.0f,0.0f,1.0f);             // Blue
            GL11.glVertex3f( 1.0f,-1.0f, 1.0f);         // Left Of Triangle (Right)
            GL11.glColor3f(0.0f,1.0f,0.0f);             // Green
            GL11.glVertex3f( 1.0f,-1.0f, -1.0f);            // Right Of Triangle (Right)
            GL11.glColor3f(1.0f,0.0f,0.0f);             // Red
            GL11.glVertex3f( 0.0f, 1.0f, 0.0f);         // Top Of Triangle (Back)
            GL11.glColor3f(0.0f,1.0f,0.0f);             // Green
            GL11.glVertex3f( 1.0f,-1.0f, -1.0f);            // Left Of Triangle (Back)
            GL11.glColor3f(0.0f,0.0f,1.0f);             // Blue
            GL11.glVertex3f(-1.0f,-1.0f, -1.0f);            // Right Of Triangle (Back)
            GL11.glColor3f(1.0f,0.0f,0.0f);             // Red
            GL11.glVertex3f( 0.0f, 1.0f, 0.0f);         // Top Of Triangle (Left)
            GL11.glColor3f(0.0f,0.0f,1.0f);             // Blue
            GL11.glVertex3f(-1.0f,-1.0f,-1.0f);         // Left Of Triangle (Left)
            GL11.glColor3f(0.0f,1.0f,0.0f);             // Green
            GL11.glVertex3f(-1.0f,-1.0f, 1.0f);         // Right Of Triangle (Left)
        GL11.glEnd();                                       // Finished Drawing The Triangle

        GL11.glLoadIdentity();                          // Reset The Current Modelview Matrix
        GL11.glTranslatef(1.5f,0.0f,-7.0f);             // Move Right 1.5 Units And Into The Screen 6.0
        GL11.glRotatef(rquad,1.0f,1.0f,1.0f);               // Rotate The Quad On The X axis ( NEW )
        GL11.glColor3f(0.5f,0.5f,1.0f);                 // Set The Color To Blue One Time Only
        GL11.glBegin(GL11.GL_QUADS);                        // Draw A Quad
            GL11.glColor3f(0.0f,1.0f,0.0f);             // Set The Color To Green
            GL11.glVertex3f( 1.0f, 1.0f,-1.0f);         // Top Right Of The Quad (Top)
            GL11.glVertex3f(-1.0f, 1.0f,-1.0f);         // Top Left Of The Quad (Top)
            GL11.glVertex3f(-1.0f, 1.0f, 1.0f);         // Bottom Left Of The Quad (Top)
            GL11.glVertex3f( 1.0f, 1.0f, 1.0f);         // Bottom Right Of The Quad (Top)
            GL11.glColor3f(1.0f,0.5f,0.0f);             // Set The Color To Orange
            GL11.glVertex3f( 1.0f,-1.0f, 1.0f);         // Top Right Of The Quad (Bottom)
            GL11.glVertex3f(-1.0f,-1.0f, 1.0f);         // Top Left Of The Quad (Bottom)
            GL11.glVertex3f(-1.0f,-1.0f,-1.0f);         // Bottom Left Of The Quad (Bottom)
            GL11.glVertex3f( 1.0f,-1.0f,-1.0f);         // Bottom Right Of The Quad (Bottom)
            GL11.glColor3f(1.0f,0.0f,0.0f);             // Set The Color To Red
            GL11.glVertex3f( 1.0f, 1.0f, 1.0f);         // Top Right Of The Quad (Front)
            GL11.glVertex3f(-1.0f, 1.0f, 1.0f);         // Top Left Of The Quad (Front)
            GL11.glVertex3f(-1.0f,-1.0f, 1.0f);         // Bottom Left Of The Quad (Front)
            GL11.glVertex3f( 1.0f,-1.0f, 1.0f);         // Bottom Right Of The Quad (Front)
            GL11.glColor3f(1.0f,1.0f,0.0f);             // Set The Color To Yellow
            GL11.glVertex3f( 1.0f,-1.0f,-1.0f);         // Bottom Left Of The Quad (Back)
            GL11.glVertex3f(-1.0f,-1.0f,-1.0f);         // Bottom Right Of The Quad (Back)
            GL11.glVertex3f(-1.0f, 1.0f,-1.0f);         // Top Right Of The Quad (Back)
            GL11.glVertex3f( 1.0f, 1.0f,-1.0f);         // Top Left Of The Quad (Back)
            GL11.glColor3f(0.0f,0.0f,1.0f);             // Set The Color To Blue
            GL11.glVertex3f(-1.0f, 1.0f, 1.0f);         // Top Right Of The Quad (Left)
            GL11.glVertex3f(-1.0f, 1.0f,-1.0f);         // Top Left Of The Quad (Left)
            GL11.glVertex3f(-1.0f,-1.0f,-1.0f);         // Bottom Left Of The Quad (Left)
            GL11.glVertex3f(-1.0f,-1.0f, 1.0f);         // Bottom Right Of The Quad (Left)
            GL11.glColor3f(1.0f,0.0f,1.0f);             // Set The Color To Violet
            GL11.glVertex3f( 1.0f, 1.0f,-1.0f);         // Top Right Of The Quad (Right)
            GL11.glVertex3f( 1.0f, 1.0f, 1.0f);         // Top Left Of The Quad (Right)
            GL11.glVertex3f( 1.0f,-1.0f, 1.0f);         // Bottom Left Of The Quad (Right)
            GL11.glVertex3f( 1.0f,-1.0f,-1.0f);         // Bottom Right Of The Quad (Right)
        GL11.glEnd();                                       // Done Drawing The Quad

        rtri+=0.2f;                                     // Increase The Rotation Variable For The Triangle ( NEW )
        rquad-=0.15f;                                   // Decrease The Rotation Variable For The Quad     ( NEW )
        return true;
    }
    private void createWindow() throws Exception {
        Display.setFullscreen(fullscreen);
        DisplayMode d[] = Display.getAvailableDisplayModes();
        for (int i = 0; i < d.length; i++) {
            if (d[i].getWidth() == 640
                && d[i].getHeight() == 480
                && d[i].getBitsPerPixel() == 32) {
                displayMode = d[i];
                break;
            }
        }
        Display.setDisplayMode(displayMode);
        Display.setTitle(windowTitle);
        Display.create();
    }
    private void init() throws Exception {
        createWindow();

        initGL();
    }

    private void initGL() {
        GL11.glEnable(GL11.GL_TEXTURE_2D); // Enable Texture Mapping
        GL11.glShadeModel(GL11.GL_SMOOTH); // Enable Smooth Shading
        GL11.glClearColor(0.0f, 0.0f, 0.0f, 0.0f); // Black Background
        GL11.glClearDepth(1.0); // Depth Buffer Setup
        GL11.glEnable(GL11.GL_DEPTH_TEST); // Enables Depth Testing
        GL11.glDepthFunc(GL11.GL_LEQUAL); // The Type Of Depth Testing To Do

        GL11.glMatrixMode(GL11.GL_PROJECTION); // Select The Projection Matrix
        GL11.glLoadIdentity(); // Reset The Projection Matrix

        // Calculate The Aspect Ratio Of The Window
        GLU.gluPerspective(
          45.0f,
          (float) displayMode.getWidth() / (float) displayMode.getHeight(),
          0.1f,
          100.0f);
        GL11.glMatrixMode(GL11.GL_MODELVIEW); // Select The Modelview Matrix

        // Really Nice Perspective Calculations
        GL11.glHint(GL11.GL_PERSPECTIVE_CORRECTION_HINT, GL11.GL_NICEST);
    }
    private static void cleanup() {
        Display.destroy();
    }
}
